### PR TITLE
fix(odp): check odp identifiers not empty before sending

### PR DESCRIPTION
**Pull request by jaeopt**
## Summary
Check if ODP events have at least one key-value pair for identifiers.
If not, discard the request without sending to the ODP server.

## Test plan
- Add a test with empty identifiers.

## Issues
- [FSSDK-8947](https://jira.sso.episerver.net/browse/FSSDK-8947)

### CODE REVIEW

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 1]
```diff
@@ -1468,6 +1468,15 @@ public ODPManager getODPManager() {
...
+
+    /**
+     * Send an event to the ODP server.
+     *
+     * @param type the event type (default = "fullstack").
+     * @param action the event action name.
+     * @param identifiers a dictionary for identifiers. The caller must provide at least one key-value pair.
+     * @param data a dictionary for associated data. The default event data will be added to this data before sending to the ODP server.
+     */
     public void sendODPEvent(@Nullable String type, @Nonnull String action, @Nullable Map<String, String> identifiers, @Nullable Map<String, Object> data) {
```
Comment (mnoman09) [inline core-api/src/main/java/com/optimizely/ab/Optimizely.java:1480]: It think we should change annotation of identifiers to Nonnull, if we are expecting that caller must provide at least one key value pair.

File: core-api/src/main/java/com/optimizely/ab/odp/ODPEventManager.java [hunk 2]
```diff
@@ -119,12 +119,19 @@ public void identifyUser(@Nullable String vuid, @Nullable String userId) {
     }
 
     public void sendEvent(ODPEvent event) {
+        event.setData(augmentCommonData(event.getData()));
```
Comment (mnoman09) [inline core-api/src/main/java/com/optimizely/ab/odp/ODPEventManager.java:122]: is there a reason why we move this before isDataValid check?

**Review by mnoman09 — CHANGES_REQUESTED** [review]
Overall looks good added few comments.

File: core-api/src/main/java/com/optimizely/ab/odp/ODPEventManager.java (hunk 2 above)
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/odp/ODPEventManager.java:122]: So, we can add extra protection against "commonData" injected from clients (android-sdk, etc) as well.

File: core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java [hunk 3]
```diff
@@ -84,4 +84,10 @@ public Boolean isDataValid() {
         }
         return true;
     }
+
+    @Transient
+    public Boolean isIdentifiersValid() {
+        return !identifiers.isEmpty();
```
[the remark this answers is not in the document]
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java:90]: This `identifiers` is instantiated in the constructor. I add Nonnull annotations to make clear.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java (hunk 1 above)
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java:1480]: @mnoman09 It turns out that this nullable option is needed to support Android (which sets non-empty common identifiers when configuring the java core). I added this clarification into the doc string. 
Another option was to add "vuid" identifier in android-sdk before passing down to java-sdk, but it was not clean since some implicit events ("identifyUser") are created in the java core.

Comment (mnoman09) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: But we are going to make this change in android as well right? Or android had to pass identifiers as null?

File: core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnoman09) [reply core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java:90]: A null check is still necessary because annotation will just give warning in some cases and null pointer exception can cause issue.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java (hunk 1 above)
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java:1480]: Android (and other client-sdks) customers are allowed to call with empty since SDK is supposed to add at least one identifiers (vuid).

File: core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/odp/ODPEvent.java:90]: @mnoman09 I may be confused, but do we still need null checking even with 
```
    @Nonnull private Map<String, String> identifiers;

     this.identifiers = identifiers != null ? identifiers : Collections.emptyMap();
```

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java (hunk 1 above)
Comment (mnoman09) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java:1480]: Sounds good.

**Review by mnoman09 — APPROVED** [review]
Lgtm
